tree.go

```diff
@@ -1075,15 +1097,76 @@ func (n *LeafNode) updateMultipleLeaves(values [][]byte) {
 	}
 }
 
-func (n *LeafNode) Delete(k []byte, _ NodeResolverFn) error {
+// Delete deletes a value from the leaf, return `true` as a second
+// return value, if the parent should entirely delete the child.
+func (n *LeafNode) Delete(k []byte, _ NodeResolverFn) (error, bool) {
 	// Sanity check: ensure the key header is the same:
 	if !equalPaths(k, n.stem) {
-		return nil
+		return nil, false
 	}
 
-	var zero [32]byte
-	n.updateLeaf(k[31], zero[:])
-	return nil
+	// Erase the value it used to contain
+	original := n.values[k[31]] // save original value
+	n.values[k[31]] = nil
+
+	// Check if a Cn subtree is entirely empty, or if
+	// the entire subtree is empty.
+	var (
+		isCnempty = true
+		isCempty  = true
+	)
+	for i := 0; i < NodeWidth; i++ {
+		if len(n.values[i]) > 0 {
+			// if i and k[31] are in the same subtree,
+			// set both values and return.
+			if byte(i/128) == k[31]/128 {
+				isCnempty = false
+				isCempty = false
+				break
+			} else {
+				// i and k[31] were in a different subtree,
+				// so all we can say at this stage, is that
+				// the whole tree isn't empty.
+				// TODO if i < 128, then k[31] >= 128 and
+				// we could skip to 128, but that's an
+				// optimization for later.
+				isCempty = false
+			}
+		}
+	}
+
+	// if the whole subtree is empty, then the
+	// entire node should be deleted.
+	if isCempty {
+		return nil, true
+	}
+
+	// if a Cn branch becomes empty as a result
+	// of removing the last value, update C by
+	// adding -Cn to it and exit.
+	if isCnempty {
+		var cn *Point
+		if k[31] < 128 {
+			cn = n.c1
+		} else {
+			cn = n.c2
+		}
+		n.commitment.Add(n.commitment, cn.Neg(cn))
+		return nil, false
+	}
+
+	// Recompute Cn', add -Cn', compute C'
+	// This is done by setting the leaf value
+	// to `nil` at this point, and all the
+	// diff computation will be performed by
+	// updateLeaf since leafToComms supports
+	// nil values.
+	// Note that the value is set to nil by
+	// the method, as it needs the original
+	// value to compute the commitment diffs.
+	n.values[k[31]] = original
+	n.updateLeaf(k[31], nil)
```

Comment:
Doesn't `updateLeaf(...)` updates commitments by using the deletion-marker?
That seems to be a deletion semantics different from all the above logic in this method, which "completely wipes out" the existence of the original value in the commitment.

Looks weird, but I could be missing context. Mentioning it just in case.

Reply:
But the Cn calulcation in the other case was clearly wrong.